sound/soc/soc-pcm.c

```diff
@@ -2789,6 +2788,114 @@ int soc_dpcm_be_digital_mute(struct snd_soc_pcm_runtime *fe, int mute)
 	return 0;
 }
 
+/*
+ * create a virtual FE DAI link
+ * Virtual FE DAI links are used in hostless pipelines
+ * to enable the codecs when the pipeline is triggered
+ */
+int soc_dpcm_vfe_new(struct snd_soc_card *card, int index,
+		     const char *link_name, const char *cpu_dai_name,
+		     const char *platform_name)
+{
+	struct snd_soc_dai_link *link;
+
+	link = kzalloc(sizeof(*link), GFP_KERNEL);
+	if (!link)
+		return -ENOMEM;
+
+	dev_dbg(card->dev, "ASoC: adding new virtual FE DAI link %s\n",
+		link_name);
+
+	/* define virtual FE DAI link */
+	link->virtual = 1;
+	link->name = link_name;
+	link->id = 1;
+	link->cpu_dai_name = cpu_dai_name;
+	link->platform_name = platform_name;
+	link->codec_name = "snd-soc-dummy";
+	link->codec_dai_name = "snd-soc-dummy-dai";
+	link->num_codecs = 1;
+
+	/* allocate memory for link codecs */
+	link->codecs = devm_kzalloc(card->dev,
+				    sizeof(struct snd_soc_dai_link_component),
+				    GFP_KERNEL);
+	if (!link->codecs)
+		return -ENOMEM;
+
+	link->codecs[0].name = link->codec_name;
+	link->codecs[0].dai_name = link->codec_dai_name;
+
+	/* enable DPCM */
+	link->dynamic = 1;
+
+	/*TODO: check if we need to handle capture for virtual FE */
+	link->dpcm_playback = 1;
```

Comment:
capture ?

Reply:
yes, this looks like a generic capability on capture where a component can capture audio and throw some sort of non-audio event DAPM knows nothing about.

Reply:
Should the virtual FE has both dpcm_playback and dpcm_capture enabled by default ?

Reply:
I've added capture support in virtual FE too. The widget creating the vfe must specify the stream direction.